src/ripple/app/tests/Path_test.cpp

```diff
@@ -180,11 +190,140 @@ find_paths(jtx::Env& env,
             std::move(sa), std::move(da));
 }
 
+Json::Value
+rpf(jtx::Account const& src, jtx::Account const& dst, std::uint32_t num_src)
+{
+    Json::Value jv = Json::objectValue;
+    jv[jss::command] = "ripple_path_find";
+    jv[jss::source_account] = toBase58(src);
+
+    if (num_src > 0)
+    {
+        auto& sc = (jv[jss::source_currencies] = Json::arrayValue);
+        Json::Value j = Json::objectValue;
+        while (num_src--)
+        {
+            j[jss::currency] = std::to_string(num_src + 100);
+            sc.append(j);
+        }
+    }
+
+    auto const d = toBase58(dst);
+    jv[jss::destination_account] = d;
+
+    Json::Value& j = (jv[jss::destination_amount] = Json::objectValue);
+    j[jss::currency] = "USD";
+    j[jss::value] = "0.01";
+    j[jss::issuer] = d;
+
+    return jv;
+}
+
 //------------------------------------------------------------------------------
 
 class Path_test : public beast::unit_test::suite
 {
 public:
+    class gate
+    {
+    private:
+        std::condition_variable cv_;
+        std::mutex mutex_;
+        bool signaled_ = false;
+
+    public:
+        // Returns `true` if signaled.
+        template <class Rep, class Period>
+        bool
+        wait_for(std::chrono::duration<Rep, Period> const& rel_time)
+        {
+            std::unique_lock<std::mutex> lk(mutex_);
+            auto b = cv_.wait_for(lk, rel_time, [=]{ return signaled_; });
+            signaled_ = false;
+            return b;
+        }
+
+        void
+        signal()
+        {
+            std::lock_guard<std::mutex> lk(mutex_);
+            signaled_ = true;
+            cv_.notify_all();
+        }
+    };
+
+    void
+    source_currencies_limit()
+    {
+        using namespace std::chrono_literals;
+        using namespace jtx;
+        testcase("source currency limits");
+        Env env(*this);
+        auto const gw = Account("gateway");
+        env.fund(XRP(10000), "alice", "bob", gw);
+        env.trust(gw["USD"](100), "alice", "bob");
+        env.close();
+
+        auto& app = env.app();
+        Resource::Charge loadType = Resource::feeReferenceRPC;
+        RPC::Context context {beast::Journal(), {}, app, loadType, app.getOPs(),
+            app.getLedgerMaster(), Role::USER, {}};
```

Comment:
Not for this PR, but the cumbersome initialization of the `Context` object could be improved with an intelligently written constructor.